### PR TITLE
Provider

### PR DESCRIPTION
**Pull request by GodReaper**

### CODE REVIEW

File: app/modules/intelligence/provider/provider_service.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+            openai_key = secret.get("api_key", os.getenv("OPENAI_API_KEY"))  # Fallback to env variable if no key
+            self.llm = ChatOpenAI(model_name="gpt-4o",api_key=openai_key, temperature=0.7, model_kwargs={"stream": True})
+        
+        elif preferred_provider == "anthropic":
+            secret = SecretManager.get_secret("anthropic", self.user_id)
+            anthropic_key = secret.get("api_key", os.getenv("ANTHROPIC_API_KEY"))  # Fallback to env variable if no key
+            self.llm = ChatAnthropic(
+                model="claude-3-5-sonnet-20240620",
+                temperature=0,
+                max_tokens=1024,
```
Comment (vineetshar) [inline app/modules/intelligence/provider/provider_service.py]: are the max tokens thought about? have you discussed with @dhirenmathur about this?

File: app/modules/intelligence/provider/provider_service.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            .first()
+        )
+        
+        # Determine preferred provider (default to 'openai')
+        preferred_provider = user_pref.preferences.get("llm_provider", "openai") if user_pref else "openai"
+        
+        if preferred_provider == "openai":
+            secret = SecretManager.get_secret("openai", self.user_id)
+            openai_key = secret.get("api_key", os.getenv("OPENAI_API_KEY"))  # Fallback to env variable if no key
+            self.llm = ChatOpenAI(model_name="gpt-4o",api_key=openai_key, temperature=0.7, model_kwargs={"stream": True})
```
Comment (vineetshar) [inline app/modules/intelligence/provider/provider_service.py]: if max tokens are present in one llm and not the other, that is inconsistent thought

File: app/modules/intelligence/provider/provider_service.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                openai_key = secret.get("api_key", os.getenv("OPENAI_API_KEY"))  # Fallback to env variable if no key
+                self.llm = ChatOpenAI(model_name="gpt-4o-mini", api_key=openai_key, temperature=0.7)
+            
+            elif preferred_provider == "anthropic":
+                secret = SecretManager.get_secret("anthropic", self.user_id)
+                anthropic_key = secret.get("api_key", os.getenv("ANTHROPIC_API_KEY"))  # Fallback to env variable if no key
+                self.llm = ChatAnthropic(
+                    model="claude-3-haiku-20240307",
+                    temperature=0,
+                    max_tokens=1024,
```
Comment (vineetshar) [inline app/modules/intelligence/provider/provider_service.py]: same thing here

File: app/modules/intelligence/provider/provider_service.py (hunk 1 above)
Comment (GodReaper) [reply app/modules/intelligence/provider/provider_service.py]: Last standup discussion for using this for agents

File: app/modules/intelligence/provider/provider_base.py [hunk 4]
```diff
@@ -0,0 +1,10 @@
+from abc import ABC, abstractmethod
+
+class ProviderBase(ABC):
```
Comment (vineetshar) [inline app/modules/intelligence/provider/provider_base.py]: why have we created this abstract class? what does it do?

File: app/modules/intelligence/provider/provider_service.py (hunk 2 above)
Comment (GodReaper) [reply app/modules/intelligence/provider/provider_service.py]: I'll take a look and do the necessary changes

File: app/modules/conversations/conversation/conversation_service.py [hunk 5]
```diff
@@ -56,59 +54,58 @@ def __init__(
...
+        llm = self.provider_service.get_llm()
+        return DebuggingAgent(llm, db)
+
+    
+    def _initialize_qna_agent(self, db: Session) -> QNAAgent:
+        llm = self.provider_service.get_llm()
+        return QNAAgent(llm, db)
+
+    # @staticmethod
+    # def _get_openai_key() -> str:
```
Comment (vineetshar) [inline app/modules/conversations/conversation/conversation_service.py]: feel free to delete the code that is not used, if there's no good reason to keep commented code. Why have we kept this one

File: app/modules/intelligence/provider/provider_service.py (hunk 1 above)
Comment (vineetshar) [reply app/modules/intelligence/provider/provider_service.py]: okay do we have the limit confirmed. Why is this specific limit ? And why not doing the same with open ai?

Comment (GodReaper) [reply app/modules/intelligence/provider/provider_service.py]: Not sure about the limit

Comment (GodReaper) [reply app/modules/intelligence/provider/provider_service.py]: I'll make it consistent

**Pull request merged**